llmfoundry/models/llm_embed/README.md

```diff
@@ -0,0 +1,184 @@
+# MPT-Embed
+
+_Q: What is a contrastive loss?_
+
+The contrastive loss can be thought of as a loss that creates a high similarity score between two similar samples, and a low score between two very different samples. Formally, this can be achieved by using the cross-entropy loss for an N-way softmax classifier. Some motivation for the contrastive loss can be found in [this blogpost](https://ankeshanand.com/blog/2020/01/26/contrative-self-supervised-learning.html). This has become the dominant method for training embedding/retrieval methods.
+
+_Q: How does the data need to be formatted?_
+
+The data simply needs to come in "pairs." This can either be in the form of:
+1. positive pairs such as "query: is snowboarding allowed at the alta ski resort?" "passage: the alta ski resort does not allow snowboarding," or
+2. tuples of positive pairs with curated hard negative pairs.
+
+In the first case, the InfoNCE Loss treats the rest of the samples in the batch as "soft negatives" (this scenario also is most successful with _very_ large global batch sizes on the order of 16-32k). In the second scenario, the InfoNCE uses the hard negatives in the denominator of the loss and much smaller global batch sizes (e.g. 32).
+
+_Q: How do you get a vector embedding out of a decoder? I thought you could only do that with encoders?_
+
+Before the final "logit" layer of the decoder, the tensor still has dimensions _batch size_ x _sequence length_ x _hidden dimension_. In order to get a single vector representation for single sample, we can simple average along the sequence length dimension (i.e. average the vectors representing each token). Alternatively, you can append an `<|endoftext|>` token to the end of each sample and extract the vector for this token alone (this seems to work well for RepLlama).
+
+The main additions are as follows:
+
+* The class `ContrastiveModel(HuggingFaceModel)`, which implements the InfoNCE Loss in the `.loss()` function. `mpt_embed` is treated as a separate model folder from `mpt`.
+* A naive dataloader for contrastive pairs `build_pairs_dataloader()`. This can handle positive pairs formatted as `text_a` and `text_b`, or positive pairs with hard negatives formatted as `query`,`passage` and a list of `hard_negatives`.
+
+## Example YAML
+
+```yaml
+name: <your run name>
+image: <the image you wish to use>
+scheduling:
+  priority: low
+  preemptible: true
+compute:
+  gpus: 16
+  cluster: r7z22
+  gpu_type: a100_40gb
+parameters:
+  seed: ${global_seed}
+  model:
```

Comment:
lots of details that don't need to be here, for documentation purposes, just show a small example model section with the contrastive name + the mpt-125m model (https://github.com/mosaicml/llm-foundry/blob/main/scripts/train/yamls/pretrain/mpt-125m.yaml), and for dataloader also just a simple yaml section. don't need the rest of the yaml (and if we were going to include a full yaml we would not want to include cluster name, wandb project names, etc.